feat: add runtime hooks for API request and response handling

mattmess1221 opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)
-->

### 🔗 Linked issue

Related to #87 and #88 
<!-- Please ensure there is an open issue and mention its number as #123 -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation, readme or JSDoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

I'm marking this as a draft because some decisions need to be made about which endpoints will trigger a hook.

Should `client -> server -> backend` send a hook on the client and server, or just the server?

<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

mattmess1221 commented:
Build failures may be related to nuxt/nuxt#29263

johannschopplich commented:
Another banger from you. 🤯 Thanks a lot, @mattmess1221!

froozeify commented:
Those hooks are what I was looking for.

I wanted to dynamically update the Authorization headers based on the request received (i.e If defined in the request as Bearer don't change it and else I'll be adding the Basic header - for an oauth app with custom oauth server)

margaretjoanmiller commented:
@froozeify did it end up working out? was going to use the hooks to set the auth token header but I'm wondering if with refresh tokens it would be more trouble than it's worth

mattmess1221 commented:
I managed to work around the build typechecking issues by importing directly from `nitropack/runtime` instead of through `#imports`. I'm thinking it may be a good idea to just avoid `#imports` altogether in the rest of the project just for consistency.

Other things I've done

- Removed hooks from server-side `$api` utility. I figure the user should explicitly pass any headers, fetch hooks, etc if they need to customize it from the server.
- Added dynamic hook keys based on the endpoint id for use of registration. e.g. `api-party:request` will trigger for all request hooks, but `api-party:request:jsonPlaceholder` will only trigger for the `jsonPlaceholder` endpoint.
